src/csharp/Microsoft.Spark/Sql/RowConstructor.cs

```diff
@@ -16,74 +18,116 @@ namespace Microsoft.Spark.Sql
         }
 
         /// <summary>
-        /// Construct a Row object from unpickled data.
-        /// Resets the cached row schema.
+        /// Construct a Row object from unpickled data. This is only to be called
+        /// on a RowConstructor that contains the row data.
         /// </summary>
         /// <returns>A row object with unpickled data</returns>
         public Row GetRow()
         {
-            // Reset the current schema to ensure that the subsequent construct()
-            // will use the correct schema if rows from different tables are being
-            // unpickled. This means that if pickled data is received per row, the
-            // JSON schema will be parsed for each row. In Spark, rows are batched
-            // such that the schema will be parsed once for a batch.
-            // One optimization is to introduce Reset() to reset the s_currentSchema,
-            // but this has a burden of remembering to call Reset() on the user side,
-            // thus not done.
-            s_currentSchema = null;
+            Debug.Assert(_parent != null);
+
+            // It is possible that an entry of a Row (row1) may itself be a Row (row2).
+            // If the entry is a RowConstructor then it will be a RowConstructor
+            // which contains the data for row2. Therefore we will call GetRow()
+            // on the RowConstructor to materialize row2 and replace the RowConstructor
+            // entry in row1.
+            for (int i = 0; i < _args.Length; ++i)
+            {
+                if (_args[i] is RowConstructor)
+                {
+                    _args[i] = ((RowConstructor)_args[i]).GetRow();
+                }
+            }
+
+            return new Row(_args, _parent.GetSchema());
+        }
+
+        /// <summary>
+        /// Clears the schema cache. Spark sends rows in batches and for each
+        /// row there is an accompany set of schemas and row entries. If the
+        /// schema was not cached, then it would need to be parsed and converted
+        /// to a StructType for every row in the batch. A new batch may contain
+        /// rows from a different table, so calling <c>Reset</c> after each
+        /// batch would aid in preventing the cache from growing too large.
+        /// Caching the schemas for each batch, ensures that each schema is
+        /// only parsed and converted to a StructType once per batch.
+        /// </summary>
+        internal void Reset()
+        {
+            s_schemaCache?.Clear();
+        }
+
+        /// <summary>
+        /// Get or cache the schema string contained in args. Calling this
+        /// is only valid if the child args contain the row values.
+        /// </summary>
+        /// <returns></returns>
+        private StructType GetSchema()
+        {
+            Debug.Assert((_args != null) && (_args.Length == 1) && (_args[0] is string));
+            string schemaString = _args[0] as string;
```

Comment:
let's do `var schemaString = (string)_args[0];` so we get casting exception instead of nullexception.

Reply:
Made suggested modifications.